### PR TITLE
Feature/lex memory reduction

### PR DESCRIPTION
**Pull request by jsmall-nvidia**
* Make Content on Token a UnownedStringSlice (before each token held 1k String !)
* Use the memory in the SourceFile if the token Content can be left as is
* Use MemoryArena on SourceManager to hold allocations for constructed strings (say ones with line continuations in them) - using (TokenFlag::ScrubbingNeeded)
* Fix places where code assumed Content held null terminated String
* Removed allocation for zero termination in CallbackWriter
* Made getFloatingPointLiteralValue and getIntegerLiteralValue return UnownedStringSlice for suffix (as opposed to String)

### CODE REVIEW

File: source/slang/lexer.cpp [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace Slang
 {
     static Token GetEndOfFileToken()
     {
-        return Token(TokenType::EndOfFile, "", SourceLoc());
+        return Token(TokenType::EndOfFile, UnownedStringSlice::fromLiteral(""), SourceLoc());
```
Comment (tfoleyNV) [inline source/slang/lexer.cpp]: Is there a reason to favor a `fromLiteral` operation here instead of just using `UnownedTerminatedStringSlice("")`?

Also: doesn't the default constructor for `UnownedStringSlice` produce an empty string anyway (just with both pointers being null)?

File: source/slang/parameter-binding.cpp [hunk 2]
```diff
@@ -999,7 +999,7 @@ static bool findLayoutArg(
     {
         if( modifier )
         {
-            *outVal = (UInt) strtoull(modifier->valToken.Content.Buffer(), nullptr, 10);
+            *outVal = (UInt) strtoull(String(modifier->valToken.Content).Buffer(), nullptr, 10);
```
Comment (tfoleyNV) [inline source/slang/parameter-binding.cpp]: Just to confirm: the conversion to `String` here is motivated by the fact that we need a termination nul byte in order to call `strtoull` and `UnownedStringSlice` doesn't guarantee that, correct?

Alternatively, this should probably be calling out to a function to extract the value of an integer literal token, rather than calling `strtoull` directly.

File: source/slang/parser.cpp [hunk 3]
```diff
@@ -990,7 +996,8 @@ namespace Slang
             case TokenType::QuestionMark:
                 if (AdvanceIf(parser, TokenType::Colon))
                 {
-                    nameToken.Content = nameToken.Content + ":";
+                    // Allocate a new slice, so we can concat : at the end
+                    nameToken.Content = parser->sink->sourceManager->allocateConcatStringSlice(nameToken.Content, UnownedStringSlice::fromLiteral(":"));
```
Comment (tfoleyNV) [inline source/slang/parser.cpp]: In our approach, operator expressions are mostly just a different syntax for function calls, so that `a + b` is treated, after parsing, just like a call `+(a,b)` to simplify the rest of the compiler (a call like that will match a function declared as `operator+`, because the function will be registered with a name of `+`).

In the case here we are parsing an expression of the form `a ? b : c`, and the `nameToken` in this case is the `?` part. Rather than construct an expression like `?(a,b,c)`, we instead construct one of the form `?:(a,b,c)` which matches a declaration of `operator?:` in the standard library. (Aside: in the case of HLSL, the `?:` operation really does act like a function, since it does not short-circuit.)

So, the adding of `:` here is really just to turn `?` into `?:`, and that is really just so that we look up the callee function based on the correct name, so it might not even be important to modify the `Content` of the token, so long as we form the correct `Name*`.

Anyway, long story short there is no reason why we have to write this as concatnation, because we should be able to just `UnownedTerminatedStringSlice("?:")`.

File: source/slang/parser.cpp [hunk 4]
```diff
@@ -2210,8 +2217,8 @@ namespace Slang
         addModifier(bufferVarDecl, reflectionNameModifier);
 
         // Both the buffer variable and its type need to have names generated
-        bufferVarDecl->nameAndLoc.name = generateName(parser, "parameterGroup_" + reflectionNameToken.Content);
-        bufferDataTypeDecl->nameAndLoc.name = generateName(parser, "ParameterGroup_" + reflectionNameToken.Content);
+        bufferVarDecl->nameAndLoc.name = generateName(parser, "parameterGroup_" + String(reflectionNameToken.Content));
```
Comment (tfoleyNV) [inline source/slang/parser.cpp]: We are *almost* at the point where we could start trying to pull out all this mess around synthesizing unique names for `cbuffer` and `tbuffer` declarations in user code.

The main reason those are needed is so that if you pass both `foo.vs.hlsl` and `foo.ps.hlsl` into Slang, and they both declare `cbuffer C { ... }` we can detect that those constant buffers are meant to represent the "same" declaration in both files (parameter binding/layout will only compute one layout for both declarations). In the end the original name is being preserved in a modifier, so we could just make the layout logic smart enough to check that instead.

This has nothing to do with this change, but I just wanted to note that the code here is due for a cleanup.

File: source/slang/parser.cpp [hunk 5]
```diff
@@ -3718,7 +3725,7 @@ namespace Slang
         {
         case TokenType::QuestionMark:
             opToken = parser->ReadToken();
-            opToken.Content = "?:";
+            opToken.Content = UnownedStringSlice::fromLiteral("?:");
```
Comment (tfoleyNV) [inline source/slang/parser.cpp]: Hmm... Here we are doing the exact same thing again, so maybe I was wrong about the earlier case of adding a `:` - one must be at the declaration site and one at the expression parsing site. They should both be able to use the simple logic that this case employs.

File: source/slang/source-loc.cpp [hunk 6]
```diff
@@ -292,6 +292,28 @@ void SourceManager::initialize(
...
+    const UInt numChars = slice.size();
+
+    char* dst = (char*)m_memoryArena.allocate(numChars);
+    ::memcpy(dst, slice.begin(), numChars);
+
+    return UnownedStringSlice(dst, numChars);
+}
+
+UnownedStringSlice SourceManager::allocateConcatStringSlice(const UnownedStringSlice& a, const UnownedStringSlice& b)
+{
```
Comment (tfoleyNV) [inline source/slang/source-loc.cpp]: If this is only being used for the `?:` case, then we should be able to eliminate this function.

File: source/slang/parameter-binding.cpp (hunk 2 above)
Comment (jsmall-nvidia) [reply source/slang/parameter-binding.cpp]: Yes - it's for null termination. Doing the conversion ourselves could avoid this, but doing correctly does require some care.

File: source/slang/parser.cpp (hunk 3 above)
Comment (jsmall-nvidia) [reply source/slang/parser.cpp]: Actually just looking at TokenType makes that clear. Will fix.

File: source/slang/parameter-binding.cpp (hunk 2 above)
Comment (tfoleyNV) [reply source/slang/parameter-binding.cpp]: String->integer is something I'm comfortable with us doing for ourselves just to make sure we are independent of any C runtime implementation differences. We may also eventually want a `BigInt` sort of type to make sure we can handle arbitrarily large literals and apply bounds checking after we know what concrete type we are expecting.

String->float (and vice versa) is something I'd like us to not do in general, although I think we currently still do our own value extraction today (issue #332 is tracking this, and links the Google double-conversion library we might want to use). String->float for hexadecimal floating-point literals is a special case that we might need/want to implement directly - its specifically the decimal case that is kind of a Bad Idea to write from scratch.

File: source/slang/lexer.cpp (hunk 1 above)
Comment (jsmall-nvidia) [reply source/slang/lexer.cpp]: On first point - because being string literal I can just get the length from the array length - 1. With UnownedTerminatedStringSlice it uses strlen. Now you could argue that the compiler will optimize that away on release, and it typically may. Moreover though I think (unless there is some special reason not to) using construction methods for a single type is better than creating new types to specially construct. Why? Because the methods and thus usage are grouped together (and easy to find), and all things being equal less types is better than more. So having UnownedStringSlice::makeTerminated(..) I think is probably superior. 

On default UnownedStringSlice, we could. We may want to distinguish between an 'empty' string and a 'never set' string or 'invalid' string. And nullptr could do that. We don't do that here, so it is arguably equivalent for our current usage.
